### PR TITLE
DrivAerML: log-cosh loss (smooth robust alternative to MSE)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Log-cosh loss = log(cosh(pred - target)) is twice differentiable everywhere, approximates L2 for small errors and L1 for large errors. Smoother than Huber (no discontinuity in derivative). May help with the DrivAerML outlier problem in a way that's more gradient-stable than Huber.

**Formula:** L(r) = log(cosh(r)) = log((exp(r) + exp(-r)) / 2)

## Instructions

Implement `--loss log_cosh` in train.py:

```python
def log_cosh_loss(pred, target):
    diff = pred - target
    return torch.mean(torch.log(torch.cosh(diff + 1e-12)))
```

Run one variant:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss log_cosh --wandb_group dm-loss-alignment
```

## Reporting Requirements
- Best val `surface_rel_l2_pct`
- Test at best val checkpoint
- W&B run ID

## Baseline

**DrivAerML baselines:**
- val best: `3.997%` @ ep467 (PR #2898, W&B: `bht6h42t`, 4L/512d, AdamW lr=5e-4, T_max=30)
- test best: `6.244%` (PR #2648, 4L/320d, truncated eval — OLDER CONFIG)
- External targets: AB-UPT 3.82% | Transolver-3: 3.71%

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Log-Cosh Loss on DrivAerML

| Metric | Baseline (MSE) | Log-Cosh | External Targets |
|--------|---------------|----------|-----------------|
| Val surface_rel_l2_pct | **3.997%** (ep 467) | 6.344% (ep 126) | — |
| Test surface_rel_l2_pct | — | 7.035% (ep 126) | AB-UPT: 3.82%, Transolver-3: 3.71% |

**Verdict: Negative result.** Log-cosh loss underperforms MSE baseline by ~59% on validation (6.34% vs 3.997%).

### Command
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss log_cosh --wandb-group dm-loss-alignment \
  --agent frieren --wandb-name "frieren/dm-log-cosh-loss"
```

### Run Details
- **W&B run ID:** z2k4uz99
- **W&B URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/z2k4uz99
- **Runtime:** ~361 minutes (21,670s), 528 epochs
- **Peak memory:** Single GPU (CUDA:0), DrivAerML default config (50k surface points)
- **Best epoch:** 126 (val 6.344%)

### What happened

1. **Fast early convergence:** Log-cosh showed notably faster initial learning — reaching 6.93% val at epoch 100 vs ~12.96% for MSE at similar epoch counts. The tanh-bounded gradient (grad = tanh(residual)) likely helps with large initial errors.

2. **Catastrophic divergence without grad-clip:** At epoch ~134, gradient norms began exploding (reaching billions), and the model entered repeated divergence-recovery cycles tied to cosine annealing LR peaks. The model never recovered its best performance.

3. **Remarkable numerical stability:** Despite grad norms exceeding 181 billion, the run never produced NaN — a direct consequence of log-cosh's tanh-bounded per-element gradients. MSE would have diverged to NaN immediately under these conditions.

4. **Why it didn't work:** Log-cosh loss produces smaller gradients than MSE for small residuals (gradient ~ residual vs 2×residual). Once the model enters a good-loss regime, the reduced gradient signal makes it harder to continue refining. Combined with cosine annealing LR peaks triggering instability, the model couldn't sustain progress past early convergence. The baseline MSE run uses grad-clip which prevents the divergence pattern entirely.

### Suggested follow-ups

- **Log-cosh + grad-clip:** The most obvious next step. Grad-clip would prevent the divergence that destroyed this run. The fast early convergence suggests log-cosh has real merit for DrivAerML's outlier-heavy distribution.
- **Adaptive log-cosh with temperature:** `log(cosh(x/τ))*τ` where τ controls the MSE-to-L1 transition point. A larger τ keeps MSE-like behavior for small residuals (better fine-tuning) while still being robust to outliers.
- **Log-cosh warmup → MSE:** Start with log-cosh for robust early convergence, then switch to MSE for precise fine-tuning.

**Comment by tcapelle** [comment]
ADVISOR: Fast early convergence with log-cosh is genuinely interesting, but the experiment diverged at ep134 because log-cosh gradients accumulate differently at cosine LR restart peaks without clipping. Please retry with **--grad-clip 1.0** added and all other settings unchanged. This is the minimum change for a clean test. Do not implement the curriculum variant yet — get the log-cosh+gc baseline first, then we can consider it.

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Log-Cosh Loss + Grad-Clip 1.0 on DrivAerML (Retry per Advisor)

| Metric | Baseline (MSE) | Log-Cosh (no GC) | Log-Cosh + GC 1.0 | External Targets |
|--------|---------------|-------------------|--------------------|-----------------|
| Val surface_rel_l2_pct | **3.997%** (ep 467) | 6.344% (ep 126) | **4.599%** (ep 287) | — |
| Test surface_rel_l2_pct | 6.244% (older cfg) | 7.035% (ep 126) | **5.672%** (ep 287) | AB-UPT: 3.82%, Transolver-3: 3.71% |

**Verdict: Improved over unclipped log-cosh, but still does not beat the MSE baseline.**

Log-cosh + grad-clip val best of 4.599% is a 27.5% improvement over unclipped log-cosh (6.344%) but still 15.1% worse than MSE baseline (3.997%).

### Command
```bash
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 \
  --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --loss log_cosh --grad-clip 1.0 --wandb-group dm-loss-alignment \
  --agent frieren --wandb-name "frieren/dm-log-cosh-gc1"
```

### Run Details
- **W&B run ID:** wnj0kf1o
- **W&B URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/wnj0kf1o
- **Runtime:** ~360 minutes (timeout), 530 epochs completed
- **Best epoch:** 287 (val 4.599%)
- **Peak memory:** Single GPU (CUDA:0), DrivAerML default config (50k surface points)

### What happened

1. **Grad-clip extended the stable training window by 3x.** The unclipped run diverged catastrophically at epoch 134. With grad-clip 1.0, training remained stable until epoch 382 — nearly three times longer. This allowed the model to continue improving well past the unclipped run's cutoff.

2. **Consistent improvement through ep311.** The val best envelope descended smoothly through each cosine cycle: 11.6% → 10.2% → 9.7% → 8.4% → 7.7% → 7.1% → 5.4% → 5.1% → 4.9% → 4.7% → 4.6%. New bests were still being set 200+ epochs past where the unclipped run died.

3. **Eventually diverged at ep382 anyway.** A cosine LR restart at ~ep379 triggered escalating grad norms (0.13 → 4.3 → 7.9 → 10.5 → 17.2 → 59K → Infinity). Grad clipping contained the explosion (no NaN ever), but model weights were corrupted. Val plateaued at ~68.7% for the remaining 150 epochs.

4. **Log-cosh's fundamental limitation persists.** The gradient signal for small residuals (tanh(r) ≈ r) is weaker than MSE (2r). This means slower fine-tuning once errors become small. The model's trough-to-trough improvement rate visibly decelerated after ep250, plateauing around 4.6-4.7%.

5. **Remarkable numerical stability.** Even with Infinity grad norms, the loss and val never produced NaN — log-cosh is bounded above by |r|, so outputs remain finite. MSE would have crashed immediately. This stability is both a feature (for exploration) and a bug (the run burned 2+ hours doing nothing useful after divergence).

### Suggested follow-ups

- **Log-cosh is not competitive with MSE on DrivAerML.** The 4.6% floor is ~15% worse than the MSE baseline (3.997%). The reduced gradient signal for small errors limits fine-tuning precision. This may close the case on log-cosh for this dataset.
- **If exploring robust losses further:** A cosine-scheduled warmdown of T_max (or switching to linear schedule) might avoid the late-stage restart instability. But the fundamental gradient-weakness issue would remain.
- **The divergence pattern is informative:** Both runs (with and without grad-clip) diverge at cosine LR restart peaks. This suggests T_max=30 may be too aggressive for log-cosh. A longer T_max (e.g., 60 or 90) would reduce peak LR spikes but also reduce regularization benefit.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3076 because ADVISOR: Closing — log-cosh loss val=4.599% doesn't beat baseline (now 3.700%). MSE + metric-aware rel-L2 auxiliary (PR #3302) is the winning loss formulation. Frieren reassigned..
